### ISSUE TITLE
Terraform destroy not deleting the resource 'google_compute_disk' intermittently

### ISSUE THREAD

**Issue by pjayaraj** [hashicorp/terraform#8667]
Terraform v0.7.0 has issues with terraform destroy with resource type 'google_compute_disk'

Please list the resources as a list, for example:
- terraform destroy fails intermittently  if it has resource of type 'google_compute_disk'
### Terraform Configuration Files

```
resource "google_compute_disk" "envprigfsnc18-glusterfs-disk-0" {
          name  = "envprigfsnc18-disk-0-${count.index}"
          type  = "pd-ssd"
          zone = "${element(split(",", var.zones), count.index)}"
          size  = "16"
          count = "3"
          lifecycle {
         prevent_destroy = false
          }
        }

  resource "google_compute_instance" "envprigfsnc18-glusterfs_server-0" {
          name = "envprigfsnc18-glusterfs-server-0-${count.index}"
          machine_type = "n1-standard-8"
          count = "3"
          zone = "${element(split(",", var.zones), count.index)}"
          metadata_startup_script="${element(data.template_file.envprigfsnc18-glusterfs-server_user_data-0.*.rendered,count.index)}"

          disk {
            image = "glusterfs-server-3-8-2-201608311147"
            size  = "90"
          }
          depends_on = ["google_compute_disk.envprigfsnc18-glusterfs-disk-0"]
          disk {
            disk = "${element(google_compute_disk.envprigfsnc18-glusterfs-disk-0.*.name, count.index)}"
            device_name = "${element(google_compute_disk.envprigfsnc18-glusterfs-disk-0.*.name, count.index)}"
            auto_delete = false
          }
          network_interface {
            subnetwork = "${var.datazone_subnet_ids}"
          }
          metadata {
           role = "glusterfs_server"
          }
          service_account {
            scopes = ["userinfo-email", "compute-ro", "storage-full"]
          }
          lifecycle {
             ignore_changes = ["metadata_startup_script"]
          }
        }

```
### Debug Output

terraform destroy
Do you really want to destroy?
  Terraform will delete all your managed infrastructure.
  There is no undo. Only 'yes' will be accepted to confirm.

  Enter a value: yes

google_compute_disk.envprigfsnc18-glusterfs-disk-1.2: Refreshing state... (ID: envprigfsnc18-disk-1-2)
data.template_file.consul_server_bootstrap_user_data: Refreshing state...
google_compute_disk.envprigfsnc18-glusterfs-disk-1.2: Destroying...
Error applying plan:

1 error(s) occurred:
- google_compute_disk.envprigfsnc18-glusterfs-disk-1.2: Error deleting disk: googleapi: Error 400: The disk resource 'envprigfsnc18-disk-1-2' is already being used by 'envprigfsnc18-glusterfs-server-1-2', resourceInUseByAnotherResource

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
### Expected Behavior

All resources should get deleted, the google_compute_instance and the google_compute_disk referred by the instance.
### Actual Behavior

Error deleting disk: googleapi: Error 400: The disk resource 'envprigfsnc18-disk-1-2' is already being used by 'envprigfsnc18-glusterfs-server-1-2', resourceInUseByAnotherResource
### Steps to Reproduce
1. `terraform apply` with a resource of type 'google_compute_disk' and a instance of type 'google_compute_instance' referring to that disk.
2.  'terraform destroy'

**Comment by paddyforan** [comment]
Hi @pjayaraj! Thanks for reporting the issue. The core issue here is that we're trying to destroyed the disk before it gets detached from the instance/before the instance is destroyed. GCP disallows that, which is the error you're seeing.

One option for us to move this forward is to detach disks explicitly before destroying them, though I don't know how feasible this is in practice. Another option is to not destroy the disks using Terraform, and just let tearing down the instance cascade into tearing down the disk. I'm not sure how feasible this is in practice, either, and it feels like it could leave state dangling.

I do believe we had a similar issue in AWS that was resolved, so I'm going to look into how that was handled, and try to emulate it here. Sorry for the problems!

**Comment by pavolculka** [comment]
hey any update on this? i am still seeing the error, is there any workaround?

**Comment by paddycarver** [comment]
Hi! sorry for the delay on this. @danawillow and I have been designing a more complete solution to some of the problems that disks are posing currently. Work on that is ongoing, but we're hopeful it will make these problems much less prevalent.

In the meantime, I think I can do a quick PR that will make this issue go away, simply be detaching the disk from any instance it's attached to when the disk is deleted. If you couple that with setting the `auto_delete` property to false when attaching the disk to the instance, Terraform's delete workflow should work as expected without any hiccups, as far as I know.

Sorry for the problems!

**Comment by paddycarver** [comment]
This should be resolved by #14651. Feel free to open another issue if this keeps cropping up, and we'll investigate!
